### PR TITLE
fix(regex): add option to disable caching regex

### LINKED ISSUE #12944: regex: add option to disable caching
### Describe the proposed change(s).

Some regex patterns use dynamic strings and should not be cached. Instead, we should:
- Add a new optional second param to `regEx` being `useCache = true`
- Add `false` as second param for all such regexes
- Ensure such regexes are moved out of loops if they *are* reused multiple times (so that we don't create them multiple times within one function call)

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes:

Added an option to disable regex caching
Implemented disable caching on some regex which are constructed using dynamic strings.

## Context:

Fixes #12944 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/config/presets/index.ts [hunk 1]
```diff
@@ -50,7 +50,7 @@ export function replaceArgs(
   if (is.string(obj)) {
     let returnStr = obj;
     for (const [arg, argVal] of Object.entries(argMapping)) {
-      const re = regEx(`{{${arg}}}`, 'g'); // TODO #12071
+      const re = regEx(`{{${arg}}}`, 'g', false); // TODO #12071
```
Comment (rarkins) [inline lib/config/presets/index.ts:53]:
```suggestion
      const re = regEx(`{{${arg}}}`, 'g', false);
```

File: lib/config/secrets.ts [hunk 2]
```diff
@@ -10,8 +10,12 @@ import { AllConfig, RenovateConfig } from './types';
...
 const secretNamePattern = '[A-Za-z][A-Za-z0-9_]*';
 
-const secretNameRegex = regEx(`^${secretNamePattern}$`);
-const secretTemplateRegex = regEx(`{{ secrets\\.(${secretNamePattern}) }}`);
+const secretNameRegex = regEx(`^${secretNamePattern}$`, undefined, false);
+const secretTemplateRegex = regEx(
+  `{{ secrets\\.(${secretNamePattern}) }}`,
+  undefined,
+  false
+);
```
Comment (rarkins) [inline lib/config/secrets.ts:18]: Revert all this. It should be cached

File: lib/datasource/go/releases-goproxy.ts [hunk 3]
```diff
@@ -107,7 +107,9 @@ export function parseNoproxy(
   }
   lexer.reset(input);
   const noproxyPattern = [...lexer].map(({ value }) => value).join('');
-  const result = noproxyPattern ? regEx(`^(?:${noproxyPattern})$`) : null;
+  const result = noproxyPattern
+    ? regEx(`^(?:${noproxyPattern})$`, undefined, false)
+    : null;
```
Comment (rarkins) [inline lib/datasource/go/releases-goproxy.ts:112]: Revert this change as this pattern is unlikely to vary much, or at all

File: lib/manager/npm/post-update/index.ts [hunk 4]
```diff
@@ -510,6 +510,7 @@ export async function getAdditionalFiles(
   } catch (err) {
     logger.warn({ err }, 'Error getting token for packageFile');
   }
+  const re = regEx(`${token}`, 'g', false);
```
Comment (rarkins) [inline lib/manager/npm/post-update/index.ts:513]:
```suggestion
  const tokenRe = regEx(`${token}`, 'g', false);
```

File: lib/manager/npm/post-update/index.ts [hunk 5]
```diff
@@ -567,7 +568,7 @@ export async function getAdditionalFiles(
         logger.debug(`${npmLock} needs updating`);
         updatedArtifacts.push({
           name: npmLock,
-          contents: res.lockFile.replace(regEx(`${token}`, 'g'), ''),
+          contents: res.lockFile.replace(re, ''),
```
Comment (rarkins) [inline lib/manager/npm/post-update/index.ts:571]:
```suggestion
          contents: res.lockFile.replace(tokenRe, ''),
```

File: lib/manager/pip_requirements/extract.ts [hunk 6]
```diff
@@ -53,8 +53,8 @@ export function extractPackageFile(
   }
   registryUrls = registryUrls.concat(extraUrls);
 
-  const pkgRegex = regEx(`^(${packagePattern})$`);
-  const pkgValRegex = regEx(`^${dependencyPattern}$`);
+  const pkgRegex = regEx(`^(${packagePattern})$`, undefined, false);
+  const pkgValRegex = regEx(`^${dependencyPattern}$`, undefined, false);
```
Comment (rarkins) [inline lib/manager/pip_requirements/extract.ts:57]: Revert

File: lib/manager/pipenv/extract.ts [hunk 7]
```diff
@@ -18,14 +18,15 @@ const specifierPartPattern = `\\s*${rangePattern.replace(
   '?:'
 )}\\s*`;
 const specifierPattern = `${specifierPartPattern}(?:,${specifierPartPattern})*`;
+const specifierRegex = regEx(`^${specifierPattern}$`, undefined, false);
```
Comment (rarkins) [inline lib/manager/pipenv/extract.ts:21]: revert

**Review by rarkins — CHANGES_REQUESTED** [review]
Some changes need reverting because they are based of module-level consts or should be cached for other reasons

File: lib/util/regex.ts [hunk 8]
```diff
@@ -19,17 +19,25 @@ try {
   RegEx = RegExp;
 }
 
-export function regEx(pattern: string | RegExp, flags?: string): RegExp {
-  const key = `${pattern.toString()}:${flags}`;
+export function regEx(
+  pattern: string | RegExp,
+  flags?: string,
```
Comment (viceice) [inline lib/util/regex.ts:24]:
```suggestion
  flags?: string | undefined,
```
Required for strict null checks

File: lib/config/secrets.ts (hunk 2 above)
Comment (viceice) [inline lib/config/secrets.ts:18]: I think this should be cached

**Comment by RahulGautamSingh** [comment]
```js
export function regEx(pattern: string | RegExp, flags?: string): RegExp {
  const key = `${pattern.toString()}:${flags}`;
 ```
Should we use a default value for flags(maybe an empty string)?
Currently the value is `undefined` incase no value is provided, which doesn't seem to be a good practice.

**Comment by rarkins** [comment]
> ```js
> export function regEx(
>   pattern: string | RegExp,
>   flags?: string | undefined,
>   useCache = true
> ): RegExp {
>   const key = `${pattern.toString()}:${flags}`; 
> ```
> 
> Should we use a default value for flags(maybe an empty string)? Currently the value is `undefined` incase no value is provided. It makes the stored key look like this `foo:undefined` which doesn't seem to be a good practice.

I think you can make the logic for constructing the key more advanced. e.g. omit the `:` if it's not defined.

**Comment by RahulGautamSingh** [comment]
> > ```js
> > export function regEx(
> >   pattern: string | RegExp,
> >   flags?: string | undefined,
> >   useCache = true
> > ): RegExp {
> >   const key = `${pattern.toString()}:${flags}`; 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Should we use a default value for flags(maybe an empty string)? Currently the value is `undefined` incase no value is provided. It makes the stored key look like this `foo:undefined` which doesn't seem to be a good practice.
> 
> I think you can make the logic for constructing the key more advanced. e.g. omit the `:` if it's not defined.

The current logic to create a key is readable and concise as it covers the 2 parts of a regex i)the pattern and ii)flags.
Changing the logic without a good reason might take the readability away I guess.
So, I have added a ternary operation to remove omit the `:` and `undefined` in absence of flag.

**Review by rarkins — APPROVED** [review]

**Pull request merged**